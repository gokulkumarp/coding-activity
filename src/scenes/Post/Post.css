.card {
  background-color: #fff;
  max-width: 800px;
  box-shadow: 2px 2px 5px #070707, -1px -1px 5px #000000;
  border-radius: 3px;
  display: grid;
  grid-template-columns: repeat(2, 1fr);
}
.card .img-container {
  width: 230px;
  height: 100%;
  grid-column: 2;
  background-color: #c1b7b4;
  background-size: cover;
  background-position: center center;
}
.card-content {
  grid-column: 3 / 5;
  padding: 10px 30px;
  border-left: 1px solid #ccc;
  text-align: justify;

}
h2 {
  text-transform: uppercase;
  color: #555;
}
h1 {
  margin-bottom: 0;
}
.card-content .author {
  border-top: 1px solid #cdcdcd;
  font-weight: 700;
  margin-top: 25px;
  padding: 25px 0 10px 0;
  color: #555;
}

@media only screen and (max-width: 600px) {
  .card {
    display: block;
  }
  .card .img-container {
    height: 250px;
    width: 100%;
  }
  .card-content {
    border: 0;
    border-top: 1px solid #ccc;
  }
}
