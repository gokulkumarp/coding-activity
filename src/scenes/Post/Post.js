import React, {useState, useEffect} from 'react';
import {nanoid} from 'nanoid';
import './Post.css';

const removeTags = (content) => {
  const regex = /(<([^>]+)>)/gi;
  return content.replace(regex, '');
}

const ErrorComponent = ({error})=>{
  return <h1>Error: {error}</h1>
}

const Post = () => {
  const [error, setError] = useState(null);
  const [isLoaded, setIsLoaded] = useState(false);
  const [item, setItem] = useState({});

  useEffect(() => {
    getData();
  }, []);

  const getData = async ()=>{
    
    try {
     const response =await fetch(
       'https://content-eu-4.content-cms.com/api/859f2008-a40a-4b92-afd0-24bb44d10124/delivery/v1/content/db4930e9-7504-4d9d-ae6c-33facca754d1'
     );
         const data = await response.json();
         setIsLoaded(true);
         setItem(data.elements);
     } catch (error) {
         setIsLoaded(true);
         setError(error);
    } 
  }

  if (error) {
    return <ErrorComponent error={error.message}/>;
  } else if (!isLoaded) {
    return <div>Loading...</div>;
  } else {
    return (
      <div>
        {item.heading ? (
          <div className="card">
            <img
              src={
                'https://content-us-1.content-cms.com/' +
                item?.mainImage.value.leadImage.url
              }
              className="img-container"
              alt={item?.mainImage.value.leadImageCaption.value}
            ></img>
            <div className="card-content">
              <h2>{item.heading.value}</h2>
              <div className="excerpt">
                {item.body.values.map((content) => (
                  <p key={nanoid()}>{removeTags(content)}</p>
                ))}
              </div>
              <p className="author">
                {item.author.value} |{' '}
                {new Intl.DateTimeFormat('en-US', {
                  year: 'numeric',
                  month: 'long',
                  day: '2-digit',
                }).format(new Date(item.date.value))}
              </p>
            </div>
          </div>
        ) : null}
      </div>
    );
  }
}

export default Post;
