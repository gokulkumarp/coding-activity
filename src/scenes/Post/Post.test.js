import { render, screen, act } from '@testing-library/react';
import Post from './Post';
import mockPostData from './mockData';

describe('Post component', () => {
    test('renders posts if request succeeds', async () => {
        window.fetch = jest.fn();
        window.fetch.mockResolvedValueOnce({
            json: async () => mockPostData
        })
        await act(async () => render(<Post />));
        const component = await screen.getByText('Top article title');
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
    });

    test('renders posts if request not succeeds', async () => {
        window.fetch = jest.fn();
        window.fetch.mockResolvedValueOnce(new Error('Internal Server Error'));
        await act(async () => render(<Post />));
        const component = await screen.getByText('Error: response.json is not a function');
        expect(component).toBeInTheDocument();
        expect(component).toMatchSnapshot();
    });
})




