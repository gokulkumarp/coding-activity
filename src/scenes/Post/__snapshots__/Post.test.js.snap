// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component renders posts if request not succeeds 1`] = `
<h1>
  Error: 
  response.json is not a function
</h1>
`;

exports[`Post component renders posts if request succeeds 1`] = `
<h2>
  Top article title
</h2>
`;
